### PR TITLE
Extract abstract box selection and add transaction building

### LINKED ISSUE #1029: Separate Box selection algorithms from ErgoBox and TrackedBox
### Motivation
Current box selection algorithms unnecessarily depend on ErgoBox and TrackedBox. 
It is not right:
1) Most of the properties are not actually used, which is not so obvious.
2) This makes the algorithms impossible to use in contexts where ErgoBox and TrackedBox are not exists (like Appkit)

### Proposal
1. Create a separate object BoxCollectors with all box collection algorithms (with layered design where one alg is based on others)
2. Define simple representation of boxes just for the sake of algorithm implementation (ERG amounts, token amounts etc.)
3. Refactor existing algorithms to use BoxCollectors

### PR DESCRIPTION
**Pull request by greenhat**
Close #1029

## Motivation
To fix https://github.com/ergoplatform/ergo-dex/issues/15 and to unify box selection and tx building across projects.
1. Use wallet's box selection in appkit (thus, in ergotool, ergodex, etc.). 
2. Add tx building code to wallet and use it in appkit (thus, in ergotool, ergodex, etc.).

## Todo:
- [x] use `TrackedBox` in `BoxSelectionResult`;
- [x] use extracted box selection algo (`BoxSelectors.select`) in wallet;
- [x] use `BoxSelectors.select` in appkit https://github.com/ergoplatform/ergo-appkit/pull/46
- [x] finish `TransactionBuild.buildUnsignedTx` implementation;
- [x] use `TransactionBuild.buildUnsignedTx` in appkit instead of it's own impl;
- [x] implement minting tokens;
- [x] implement all appropriate checks from `ErgoTransaction.validateStateless`;  
**Required:**
- [x] make `BoxSelection` hierarchy polymorphic in box type;
- [x] switch from `Option` to `Either` for box selection result to expose errors;
- [x] test in appkit and dex;
- [x] re-asses box selection algos (see review notes); 
**Nice to have:**
- [ ] implement all appropriate checks from `ErgoTransaction.validateStatefull`;
- [ ] add more tests for tx builder;
- [ ] pass min box value rather instead of hard-coded `BoxSelector.MinBoxValue`;

## Notes to a reviewer
1. Box selection  algo's from `BoxSelector` and `DefaultBoxSelector` extracted into `BoxSelectors`.
2. `ErgoBoxAssets` trait introduced in sigma to expose the only needed info (box'es value and tokens) needed to perform box selection. PR https://github.com/ScorexFoundation/sigmastate-interpreter/pull/646
3. Introduced `TransactionBuilder.buildUnsignedTx` that add miner's fee box and change box(es) along with performing appropriate checks ensuring that resulted tx will be accepted by the node.  **I'm calling box selection inside tx building** passing the inputs to get the change boxes. Although it does not look very nice it should do the job (also I'm checking that selected boxes = input boxes).
4. **Hard-coded `BoxSelector.MinBoxValue` that is used as a threshold for change box value. Box selection fails if ERG value for change box < `BoxSelector.MinBoxValue`.**

## Downstream PRs:
appkit - https://github.com/ergoplatform/ergo-appkit/pull/46
dex - https://github.com/ergoplatform/ergo-dex/pull/14
tool - https://github.com/ergoplatform/ergo-tool/pull/53

### CODE REVIEW

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/BoxSelector.scala [hunk 1]
```diff
@@ -1,10 +1,11 @@
 package org.ergoplatform.wallet.boxes
 
 import org.ergoplatform.ErgoBox
-import org.ergoplatform.wallet.boxes.BoxSelector.{BoxSelectionResult, subtractAssetsMut}
+import org.ergoplatform.wallet.boxes.BoxSelector.{BoxSelectionResult}//, subtractAssetsMut}
```
[the remark this answers is not in the document]
Comment (greenhat) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/BoxSelector.scala]: Already removed by @kushti. :)

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/BoxSelectors.scala [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+      case (id, amount) =>
+        into.put(id, into.getOrElse(id, 0L) + amount)
+    })
+  }
+
+  // def mergeAssets(from: Map[ModifierId, Long], to: Map[ModifierId, Long] = Map.empty): Map[ModifierId, Long] = {
+  //   from.foldLeft(to) { case (acc, (id, amount)) =>
+  //     acc.updated(id, acc.getOrElse(id, 0L) + amount)
+  //   }
+  // }
```
[the remark this answers is not in the document]
Comment (greenhat) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/BoxSelectors.scala]: Done.

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/BoxSelectors.scala [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+      compactedAssets.grouped(ErgoBox.MaxTokens).toSeq
+    val changeBalance = compactedBalance - targetBalance
+    formChangeBoxes(changeBalance, changeBoxesAssets)
+  }
+
+  def formChangeBoxes(
+    changeBalance: Long,
+    changeBoxesAssets: Seq[mutable.Map[ModifierId, Long]]
+  ): Option[Seq[ErgoBoxAssets]] = {
+    //at least 1 ergo token should be assigned per a created box
```
[the remark this answers is not in the document]
Comment (greenhat) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/BoxSelectors.scala]: Yeah, it caught my eye as well. Maybe at that time, we did not have minimum amount of ergo per box. 
My initial plan was to just extract the existing algo and bring it to ergodex ASAP, so I did not want to change it. Now, seems like a good time to re-assess the algo.

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/TrackedBox.scala [hunk 4]
```diff
@@ -62,14 +63,14 @@ case class TrackedBox(creationTxId: ModifierId,
...
+  override def value: Long = box.value
 
-  def assets: Map[ModifierId, Long] = box.additionalTokens.toArray.map {
+  override def tokens: Map[ModifierId, Long] = box.additionalTokens.toArray.map {
     case (id, amt) => bytesToId(id) -> amt
   }.toMap
 
   override def equals(obj: Any): Boolean = obj match {
-    case tb: TrackedBox => tb.creationTxId == creationTxId && tb.creationOutIndex == creationOutIndex
+    case tb: TrackedBox => tb.creationTxId == creationTxId && tb.creationOutIndex == creationOutIndex && box == tb.box
```
[the remark this answers is not in the document]
Comment (greenhat) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/TrackedBox.scala]: I hit it in a test where TrackedBox'es were generated with the same tx id and index. Although the test data were incorrect, it got me thinking that existing equality was too weak, so I added the box itself.

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/transactions/TransactionBuilder.scala [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+object TransactionBuilder {
+
+  private def calcTokenOutput(outputCandidates: Seq[ErgoBoxCandidate]): Map[ModifierId, Long] = 
+    outputCandidates
+      .map(b => collTokensToMap(b.additionalTokens))
+      .foldLeft(Map[ModifierId, Long]()){case (a, e) => a.combine(e) }
+
+  // TODO: extract into an Iso
+  private def collTokensToMap(tokens: Coll[(TokenId, Long)]): Map[ModifierId, Long] = 
+    tokens.toArray.toSeq.map(t => bytesToId(t._1) -> t._2).toMap
```
[the remark this answers is not in the document]
Comment (greenhat) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/transactions/TransactionBuilder.scala]: Yes, you're right. I missed that.

**Comment by greenhat** [comment]
Progress so far:
– made `BoxSelector` hierarchy polymorphic in box type;
- ditched `BoxSelectors` in favor of `DefaultBoxSelector`;
- switched from `Option` to `Either` for box selections results to convey the exact error;

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/BoxSelectors.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (greenhat) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/BoxSelectors.scala]: Hard-coded 1000000 in `BoxSelector.MinBoxValue` for now, with plans to make it a parameter later.

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/BoxSelectors.scala [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    if (changeBoxesAssets.size > changeBalance) {
+      None
+    } else {
+      val changeBoxes = if (changeBoxesAssets.nonEmpty) {
+        val baseChangeBalance = changeBalance / changeBoxesAssets.size
```
[the remark this answers is not in the document]
Comment (greenhat) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/BoxSelectors.scala]: Fixed in the check above.

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/BoxSelectors.scala [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+  //   }
+  // }
+
+  private def subtractAssetsMut(
+    from: mutable.Map[ModifierId, Long],
+    subtractor: Map[ModifierId, Long]
+  ): Unit = {
+    subtractor.foreach {
+      case (id, subtractAmt) =>
+        val fromAmt = from(id)
```
[the remark this answers is not in the document]
Comment (greenhat) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/BoxSelectors.scala]: Guarded with `getOrElse`.

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/BoxSelectors.scala [hunk 8]
```diff
@@ -0,0 +1,170 @@
+package org.ergoplatform.wallet.boxes
+
+import scorex.util.ModifierId
+import scala.annotation.tailrec
+import scala.collection.mutable
+import org.ergoplatform.ErgoBox
+import org.ergoplatform.ErgoBoxAssets
+import org.ergoplatform.ErgoBoxAssetsHolder
+
+object BoxSelectors {
```
[the remark this answers is not in the document]
Comment (greenhat) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/BoxSelectors.scala]: I moved the code back to `BoxSelector` hierarchy and deleted `BoxSelectors`.

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/BoxSelectors.scala [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+                  targetAmt > 0 && targetAmt > currentAmt
+              },
+            assetsMet
+          )) {
+        subtractAssetsMut(currentAssets, targetAssets)
+        val changeBoxesAssets: Seq[mutable.Map[ModifierId, Long]] =
+          currentAssets.grouped(ErgoBox.MaxTokens).toSeq
+        val changeBalance = currentBalance - targetBalance
+        formChangeBoxes(changeBalance, changeBoxesAssets).map(changeBoxes =>
+          SelectionResult(res, changeBoxes)
```
[the remark this answers is not in the document]
Comment (greenhat) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/BoxSelectors.scala]: Fixed

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/Utils.scala [hunk 10]
```diff
@@ -64,4 +64,16 @@ object Utils {
     )
   }
 
+  implicit class EitherOpsFor211[+A, +B](val source: Either[A, B]) extends AnyVal {
```
Comment (aslesarenko) [inline ergo-wallet/src/main/scala/org/ergoplatform/wallet/Utils.scala]: @greenhat Please add ScalaDoc with more create description of this.

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/transactions/TransactionBuilder.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/transactions/TransactionBuilder.scala]: Why not use LinkedListMap then? We are hiding subtle potential bugs here

File: ergo-wallet/src/test/scala/org/ergoplatform/wallet/transactions/TransactionBuilderSpec.scala [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+  implicit val addressEncoder = new ErgoAddressEncoder(
+    ErgoAddressEncoder.TestnetNetworkPrefix
+  )
+
+  val seedStr                       = "edge talent poet tortoise trumpet dose"
+  val seed: Array[Byte]             = Mnemonic.toSeed(seedStr)
+  val rootSecret: ExtendedSecretKey = ExtendedSecretKey.deriveMasterKey(seed)
+
+  val currentHeight    = 0
+  val minFee           = 1000000L
```
Comment (aslesarenko) [inline ergo-wallet/src/test/scala/org/ergoplatform/wallet/transactions/TransactionBuilderSpec.scala]: shouldn't it be the actual MinFee?

File: ergo-wallet/src/test/scala/org/ergoplatform/wallet/transactions/TransactionBuilderSpec.scala [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+  val minChangeValue   = 1000000L
```
Comment (aslesarenko) [inline ergo-wallet/src/test/scala/org/ergoplatform/wallet/transactions/TransactionBuilderSpec.scala]: Also, why not use actual MinFee?

File: ergo-wallet/src/test/scala/org/ergoplatform/wallet/transactions/TransactionBuilderSpec.scala [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+  val minerRewardDelay = 720
+
+  val TrueLeaf: SigmaPropValue = Values.TrueLeaf.toSigmaProp
```
Comment (aslesarenko) [inline ergo-wallet/src/test/scala/org/ergoplatform/wallet/transactions/TransactionBuilderSpec.scala]: When it is SigmaProp, why not call it TrueProp?

**Comment by greenhat** [comment]
@kushti Good catch! Moved box selector to parameters.

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/Utils.scala (hunk 10 above)
Comment (greenhat) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/Utils.scala]: Done.

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/transactions/TransactionBuilder.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (greenhat) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/transactions/TransactionBuilder.scala]: Order preservation is not required here. If the reason is to make iso out of it later, I'd rather not spend time on it now.

File: ergo-wallet/src/test/scala/org/ergoplatform/wallet/transactions/TransactionBuilderSpec.scala (hunk 11 above)
Comment (greenhat) [reply ergo-wallet/src/test/scala/org/ergoplatform/wallet/transactions/TransactionBuilderSpec.scala]: Thanks! Fixed.

File: ergo-wallet/src/test/scala/org/ergoplatform/wallet/transactions/TransactionBuilderSpec.scala (hunk 12 above)
Comment (greenhat) [reply ergo-wallet/src/test/scala/org/ergoplatform/wallet/transactions/TransactionBuilderSpec.scala]: Fixed.

File: ergo-wallet/src/test/scala/org/ergoplatform/wallet/transactions/TransactionBuilderSpec.scala (hunk 13 above)
Comment (greenhat) [reply ergo-wallet/src/test/scala/org/ergoplatform/wallet/transactions/TransactionBuilderSpec.scala]: Done.

**Pull request merged**